colossalai/communication/profiler.py

```diff
@@ -0,0 +1,262 @@
+              "total number of calls: {}".format(self.total_count),
+              "All events:", sep='\n')
+
+        show_list = sorted(self.ops_record.items(), key=lambda kv: -kv[1].self_cuda_time)
+        for location, event in show_list:
+            print(location,
+                  "self cuda time: {}".format(_format_time(event.self_cuda_time)),
+                  "{:.1f}% of total communication time".format(event.self_cuda_time / self.total_cuda_time * 100.0),
+                  "average bandwith: {}".format(_format_bandwith(event.self_comm_vol, event.self_cuda_time)),
+                  "number of calls: {}".format(event.self_count),
+                  "--------------------", sep='\n')
+
+    @property
+    def has_aync_op(self):
+        return self.pending_op is not None
+
+    def activate_profiler(self, enable: bool, kn: str, vol: float):
+        self.pending_metadata = (kn, _get_code_location(3), vol)
+        self.profiler = profile(enabled=enable, use_cuda=True,
+                                use_cpu=True, use_kineto=True)
+        self.profiler.__enter__()
+
+    def close_profiler(self):
+        assert self.profiler is not None, "There is no running dist op"
+        kernel_name, code_location, vol = self.pending_metadata
+        self.profiler.__exit__(None, None, None)
+
+        if self.profiler.enabled:
+            assert_flag = 0
+            current_comm_event = None
+            events = self.profiler.function_events
+            for event in events:
+                if kernel_name in event.name:
+                    assert assert_flag == 0, "Multiple dist ops has been called "
+                    current_comm_event = CommEvent(1, vol, event.self_cuda_time_total)
+                    assert_flag += 1
+
+            assert current_comm_event is not None, "dist op has not been found"
+            self.total_count += current_comm_event.self_count
+            self.total_comm_vol += current_comm_event.self_comm_vol
+            self.total_cuda_time += current_comm_event.self_cuda_time
+            if code_location in self.ops_record:
+                self.ops_record[code_location].add(current_comm_event)
+            else:
+                self.ops_record[code_location] = current_comm_event
+
+        self.profiler = None
+        self.pending_op = None
+        self.pending_metadata = None
+
+    def wait_async_op(self):
+        if self.pending_op is not None:
+            op = self.pending_op
+            op.wait()
+            self.close_profiler()
+
+
+CCPROF = CommProf()
+
+
+def cc_wait_aync():
+    CCPROF.wait_async_op()
+
+
+def cc_all_reduce(tensor: torch.Tensor,
+                  op: ReduceOp = ReduceOp.SUM,
+                  pm: ParallelMode = ParallelMode.GLOBAL,
+                  async_op: bool = False,
+                  enable_profile: bool = False) -> None:
+    comm_size = gpc.get_world_size(pm)
+
+    if comm_size == 1:
+        return
+
+    CCPROF.wait_async_op()

```

Comment:
There are some duplicated code in different function.
Can you implement it as a wrapper?

Reply:
https://www.geeksforgeeks.org/function-wrappers-in-python/